src/sleep/sleep.c

```diff
@@ -300,37 +305,54 @@ static int execute_s2h(const SleepConfig *sleep_config) {
 
-                        log_debug("%d%% battery discharged in %s time",
+                        log_debug("%d%% battery discharged in %f seconds time",
                                         last_capacity - current_capacity,
-                                        FORMAT_TIMESPAN(after_timestamp - before_timestamp, USEC_PER_SEC));
-                        estimated_total_discharge_time = (current_capacity * (before_timestamp - after_timestamp)) / (last_capacity - current_capacity);
-                        log_debug("Battery will discharge in %s time", FORMAT_TIMESPAN(estimated_total_discharge_time, USEC_PER_SEC));
+                                        time_elapsed);
+
+                        estimated_discharge_rate = ((last_capacity - current_capacity) * 60 * 60) / time_elapsed;
+
+                        log_debug("Battery discharge rate is %d%% per hour", estimated_discharge_rate);
+
+                        if (estimated_discharge_rate > 0 && estimated_discharge_rate < 1000) {
+                                r = put_battery_discharge_rate(estimated_discharge_rate);
+                                if (r < 0)
+                                        return log_warning_errno(r, "Failed to update battery discharge rate: ignoring %m");
+                        }
                         return 0;
                 }
+
                 tfd = safe_close(tfd);
 
+                log_debug("Attempting to estimate battery after waking from %s hours sleep timer",
+                          FORMAT_TIMESPAN(suspend_interval, USEC_PER_HOUR));
+
                 /* If woken up after alarm time, calculate battery capacity and battery capacity_level and estimate discharge rate */
+                current_capacity = read_battery_capacity_percentage();
 
-                current_capacity = read_battery_capacity();
                 if (current_capacity < 0)
                         return log_error_errno(errno, "Error fetching battery capacity: %m");
 
-                estimated_total_discharge_time = current_capacity / (last_capacity - current_capacity);
+                estimated_discharge_rate = (last_capacity - current_capacity) / (suspend_interval * USEC_PER_HOUR);
 
-                log_debug("Battery will discharge in %f hours", estimated_total_discharge_time);
-
-                suspend_interval = ((estimated_total_discharge_time * 60) - 30) * USEC_PER_MINUTE;
-                /* 30min wiggle time. Finally suspend_interval is stored in minutes*/
+                log_debug("Battery discharge rate is %d%% per hour", estimated_discharge_rate);
+                /* log debug to be removed later */
+                if (estimated_discharge_rate > 0 && estimated_discharge_rate < 1000) {
+                        r = put_battery_discharge_rate(estimated_discharge_rate);
+                        if (r < 0)
+                                return log_warning_errno(r, "Failed to update battery discharge rate: ignoring %m");
```

Comment:
this should not be fatal, it can continue with best-effort and provide the core functionality, so log and ignore